staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,551 @@
+			// Configuration error.
+			addConfigError(definitionInfo.configurationError, definition, nil)
+			continue
+		}
+
+		dependentBindings := c.definitionsToBindings[definitionNamespacedName]
+		if len(dependentBindings) == 0 {
+			// Definition has no known bindings yet.
+			addConfigError(errors.New("no bindings found"), definition, nil)
+			continue
+		}
+
+		for namespacedBindingName := range dependentBindings {
+			// If the key is inside dependentBindings, there is guaranteed to
+			// be a bindingInfo for it
+			bindingInfo := c.bindingInfos[namespacedBindingName]
+			binding := bindingInfo.lastReconciledValue
+			if !binding.Matches(a) {
+				continue
+			}
+
+			paramsNamespace, paramsName := binding.GetTargetParams()
+			// Find the params referred by the binding by looking its name up
+			// in our informer for its CRD
+			paramInfo, ok := c.paramsCRDControllers[definition.GetParamSource()]
+			if !ok {
+				addConfigError(fmt.Errorf("paramSource kind `%v` not known",
+					definition.GetParamSource().String()), definition, binding)
+				continue
+			}
+
+			var param *unstructured.Unstructured
+			if len(paramsNamespace) == 0 {
+				param, err = paramInfo.controller.Informer().Get(paramsName)
+			} else {
+				param, err = paramInfo.controller.Informer().Namespaced(paramsNamespace).Get(paramsName)
+			}
+
+			if err != nil {
+				// Apply failure policy
+				addConfigError(err, definition, binding)
+
+				if k8serrors.IsNotFound(err) {
+					// Param doesnt exist yet?
+					// Maybe just have to wait a bit.
+					continue
+				}
+
+				// There was a bad internal error
+				utilruntime.HandleError(err)
+				continue
+			}
+
+			if bindingInfo.evaluator == nil {
+				// Compile policy definition using binding
+				bindingInfo.evaluator, err = definition.Compile(c.objectConverter)
+				if err != nil {
+					// compilation error. Apply failure policy
+					wrappedError := fmt.Errorf("failed to compile CEL expression: %w", err)
+					addConfigError(wrappedError, definition, binding)
+					continue
+				}
+				c.bindingInfos[namespacedBindingName] = bindingInfo
+			}
+
+			decisions := bindingInfo.evaluator(a, param)
+			for _, decision := range decisions {
+				switch decision.PolicyDecisionKind {
+				case Admit:
+					// Do nothing
+				case Deny:
+					allDecisions = append(allDecisions, PolicyDecisionWithMetadata{
+						Definition:     definition,
+						Binding:        binding,
+						PolicyDecision: decision,
+					})
+				default:
+					// unrecognized decision. ignore
+				}
+			}
+		}
+	}
+
+	if len(allDecisions) > 0 {
+		return k8serrors.NewConflict(
+			a.GetResource().GroupResource(), a.GetName(),
+			PolicyError{
+				Decisions: allDecisions,
+			})
+	}
+
+	return nil
+}
+func (c *celAdmissionController) HasSynced() bool {
+	return c.policyBindingController.HasSynced() &&
+		c.policyDefinitionsController.HasSynced()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Test Helpers
```

Comment:
Can you move these to the test file, or otherwise refactor so that production and test code aren't mixed?